### ISSUE TITLE
Deadlock when Running CCLs on Galaxy

### ISSUE THREAD

**Issue by tt-asaigal** [tenstorrent/tt-metal#10978]
### Problem

When running a CCL on multiple chips on the same tunnel (currently only supported for TG and TGG), it is possible to enter a deadlock due to Demux being a shared resource (has 1 input queue shared across `prefetch_h`).

Example:
Since commands can be sent to devices out of order, on 2 devices belonging to the same tunnel (Tunnel Stop 1 and Tunnel Stop 2), the following code:

```
input_tensor.to(device_mesh) # Send tensor to a device mesh, that has devices on the same tunnel
output = ttnn.line_all_gather(input_tensor, dim, num_links=num_links, memory_config=mem_config)
input_tensor2.to(device_mesh)
```

Can get executed as:

```
1. Tunnel Stop 1 and 2 get EnqueueWriteBuffer commands to place input_tensor in device memory
2. prefetch_h for Tunnel Stop 1 sends commands for the CCL over Mux/Demux
3. dispatch_d for Tunnel Stop 1 waits for workers running the CCL to complete (this requires Tunnel Stop 2 to at least start 
    executing the CCL)
4. (Out of Order) prefetch_h for Tunnel Stop 1 sends EnqueueWriteBuffer commands to place input_tensor2 in device memory.
     prefetch_d sends commands to dispatch_d for this step. At some point dispatch_d back-pressures, since it's stalled and can't 
     release pages.
     prefetch_d stalls trying to acquire pages for the dispatch_d CB. At some point prefetch_d back-pressures (if the second write is 
     large).
     Demux on Tunnel Stop 1 has a valid packet to send to prefetch_d, but gets stalled, since the cmddat_q for prefetch_d is full.
5. (Out of Order) prefetch_h for Tunnel stop 2 sends commands for the CCL over Mux/Demux.
     Demux is still stalled, so packets for this command cannot be forwarded, and the CCL on this chip never starts.
```

For the CCL to complete, Tunnel Stop 2 must receive the `EnqueueProgram` command.

### Proposal 

The long term solution for this is to use virtual channels (one channel per chip, with demux serving each channel independently), so that the traffic can be decoupled.

Until VCs are implemented, we will tag CCLs as special OPs, that require a pipeline-flush, once they have been sent to `prefetch_h`.

This flush will be implemented using events:

After `EnqueueProgram` is called for CCL, a `record_event` command will be sent to `dispatch_d` on the chip running this Op. The corresponding `prefetch_h` core will get a `wait_for_event` signal, i.e.: it will only send additional data through the packet network once `dispatch_d` notifies it of CCL completion.

During CCL execution, the pipeline will be empty, since `prefetch_h` is stalled.

This flush is currently only needed for Galaxy systems, since they have deep tunnels and will require minor changes to CCL ops:
Programs need to be tagged using the `program.capture_multi_device_dependencies()` API.

**Comment by tt-asaigal** [comment]
@SeanNijjar @cfjchu This issue has been resolved in https://github.com/tenstorrent/tt-metal/pull/10875. Please take a look at the minor change to the line all-gather op, which bypasses this hang.

Until VCs are supported, all CCLs running on Galaxy will need to tag their programs in a similar manner.

fyi @pgkeller @davorchap

**Comment by tt-asaigal** [comment]
Reduce Scatter support staged at https://github.com/tenstorrent/tt-metal/pull/11077.

**Comment by ubcheema** [comment]
Fix has been pushed to main @ 66cb06dfdafcc5e3e9d37e8d9d49170773e0eb60
Issue: https://github.com/tenstorrent/tt-metal/issues/12044

**Comment by tt-asaigal** [comment]
VC Support is in, closing,
